augur/distance.py

```diff
@@ -22,6 +24,296 @@ def read_masks(mask_file):
+        # If the given latest date is not None, skip nodes that were sampled
+        # prior to this date.
+        if latest_date is not None and node.attr["num_date"] < latest_date:
+            continue
+
+        # Find the closest ancestor of this node that was also sampled prior to
+        # the given latest date. Stop searching once we reach the root. If the
+        # latest date requested is None, the immediate parent of each node will
+        # be used.
+        parent = node.parent
+        while parent != tree.root and latest_date is not None and parent.attr["num_date"] > latest_date:
+            parent = parent.parent
+
+        # Calculate distance between current node and its ancestor.
+        distances_by_node[node.name] = get_distance_between_nodes(
+            sequences_by_node_and_gene[parent.name],
+            sequences_by_node_and_gene[node.name],
+            distance_map
+        )
+
+    return distances_by_node
+
+
+def get_distances_to_all_pairs(tree, sequences_by_node_and_gene, distance_map, earliest_date=None, latest_date=None):
+    """Calculate distances between each sample in the given sequences and all other
+    samples in previous seasons using the given distance map.
+
+    Parameters
+    ----------
+    tree : Bio.Phylo
+        a rooted tree whose node names match the given dictionary of sequences
+        by node and gene
+
+    sequences_by_node_and_gene : dict
+        nucleotide or amino acid sequences by node name and gene
+
+    distance_map : dict
+        site-specific and, optionally, sequence-specific distances between two
+        sequences
+
+    earliest_date, latest_date : pandas.Timestamp
+        earliest or latest date to consider a node for comparison to a given
+        sample; used to define a range of previous seasons relative to the most
+        recent samples in the given tree.
+
+    Returns
+    -------
+    dict :
+        distances calculated between each sample in the tree and all samples
+        from previous samples with distances indexed by primary sample name and
+        then past sample name
+
+    """
+    if earliest_date is not None:
+        earliest_date = timestamp_to_float(earliest_date)
+
+    if latest_date is not None:
+        latest_date = timestamp_to_float(latest_date)
+
+    distances_by_node = {}
+
+    # Calculate distance between each tip and all tips in previous seasons as
+    # defined by the given latest date threshold.
+    for node in tree.find_clades(terminal=True):
+        # Skip nodes that were sampled prior to this date.
+        if latest_date is not None and node.attr["num_date"] < latest_date:
+            continue
+
+        # Distances between this node and other nodes are indexed by the other
+        # node name.
+        distances_by_node[node.name] = {}
+
+        # Find all nodes that were sampled prior to the given latest date.
+        for past_node in tree.find_clades(terminal=True):
```

Comment:
Should a node's distance to itself be omitted, e.g. when `node.name == past_node.name`?